design/baremetal-operator/bmh_boot_iso.md

```diff
@@ -0,0 +1,128 @@
+for example the [fedora-coreos installer](https://docs.fedoraproject.org/en-US/fedora-coreos/bare-metal/)
+
+An [Ironic API](https://docs.openstack.org/ironic/latest/admin/drivers/redfish.html#virtual-media-ramdisk) exists for this, but it's not currently accessible via metal3, so it would be desirable to optionally enable this workflow.
+
+### Goals
+
+Expose the Ironic API to boot iso images via metal3.
+
+### Non-Goals
+
+This only considers the case where the redfish-virtualmedia BMC
+option is used, not any ability to boot arbitrary iso images via
+PXE.
+
+## Proposal
+
+Add a new field to the BareMetalHost spec which indicates that instead of
+deploying an image to disk, a live-iso image will be booted:
+
+```
+  spec:
+    image:
+      iso: http://1.2.3.4/image.iso
+    online: true
+```
+
+Note that in this mode, `rootDeviceHints` and `userData` will not be used
+since the image won't be written to disk, and Ironic doesn't
+support passing user-data in addition to the iso attachment at present.
+
+At some future point it would be desirable for Ironic to add the ability
+to pass userData - then a generic iso could be booted with the ability
+to define customization via the userData field in the normal way.
+
+## Design Details
+
+When this mode is selected, we need to configure Ironic to use the [ramdisk
+deploy interface](https://docs.openstack.org/ironic/stein/admin/interfaces/deploy.html#ramdisk-deploy)
+which is not currently enabled in ironic-image.
+
+We will also need to write the provided ISO URL into the
+[instance_info boot_iso field](https://docs.openstack.org/ironic/latest/admin/drivers/redfish.html#virtual-media-ramdisk)
+
+This will only work when the redfish-virtual-media boot interface is selected
+which will be the case when the `redfish-virtualmedia+http://foo` bmc address
+is specified - we need to ensure validation exists to catch invalid combinations
+
+### Risks and Mitigations
+
+- Currently there is no detachment API in ironic, so in the case where an installer
+iso is booted, we don't have a way to detach the virtualmedia and will have to
+rely e.g on efibootmgr in whatever iso gets booted to ensure the correct boot device.
+
+- The inspection of a BMH will still rely on booting IPA and there's no support for
+fast-track provisioning in this workflow (since it's booting two different iso images)
+so this adds a reboot into the process.  This can potentially be avoided in some cases
+where we can provide a status annotation on creation of the BMH such that inspection
+is not performed.
+
+- Similarly cleaning of a BMH will rely on IPA, so will require a reboot which may
+not be desirable in some situations.  This can potentially be avoided if cleaning
+is globally disabled in the ironic-image configuration via the
+[IRONIC_AUTOMATED_CLEAN](https://github.com/metal3-io/ironic-image/blob/master/configure-ironic.sh#L39)
+variable but this doesn't provide any per-host control.
```

Comment:
We have had other people ask for a way to disable cleaning on deprovisioning. I think we can consider dealing with that via a separate design proposal.

Reply:
@fmuyassarov is working on a proposal for that already

Reply:
Yes, I will open a proposal in metal3-docs for having a configurable disk cleaning, latest end of this week.

Reply:
I think I saw a comment that we're cleaning on both provisioning (the ironic `provide` state transition) and deprovisioning, is the plan to either remove one or make both configurable?

My concern is that for this use-case, we want to boot the FCOS installer iso as fast as possible, and there is already code in the image to handle wiping disks etc.

Reply:
The plan is to be able to completely disable/enable disk cleaning from metal3 per node. This requires some changes in the Ironic code as well, because, once we enable disk cleaning for a node, there is no way back to disable it again.

So, currently if we disable automated cleaning globally, then we can enable it only for certain nodes that require disk cleaning as part of deprovisioning. However, from the security perspective the safer option would be to enable automated cleaning globally and then disable it for certain nodes that don't require disk cleaning. And this option is not supported currently. 

We had quick chat on: https://github.com/metal3-io/baremetal-operator/issues/626 and the issue is opened in Ironic https://storyboard.openstack.org/#!/story/2008113.

Reply:
A patch is on review for this feature.

Reply:
Added a link to https://github.com/metal3-io/metal3-docs/pull/151